test/test_wave.py

```diff
@@ -98,165 +98,104 @@ def _cos(self, val):
+    [
+        get_standing_wave(2),
+        get_standing_wave(3),
+        get_manufactured_cubic(2),
+        get_manufactured_cubic(3)
+    ])
 @pytest.mark.parametrize("order", [2, 3, 4])
-def test_wave_manufactured(ctx_factory, dim, order):
+def test_wave(ctx_factory, dim, order, c, mesh_factory, sym_phi):
     cl_ctx = ctx_factory()
     queue = cl.CommandQueue(cl_ctx)
 
     from pytools.convergence import EOCRecorder
     eoc_rec = EOCRecorder()
 
     for n in [4, 8, 16]:
-        from meshmode.mesh.generation import generate_regular_rect_mesh
-        mesh = generate_regular_rect_mesh(
-                a=(-1.,)*dim,
-                b=(1.,)*dim,
-                n=(n,)*dim)
+        mesh = mesh_factory(n)
 
         from grudge.eager import EagerDGDiscretization
         discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
 
         nodes = discr.nodes().with_queue(queue)
 
-        # 2D: phi(x,y,t) = cos(omega*t-pi/4)*(x-1)^3*(x+1)^3*(y-1)^3*(y+1)^3
-        # 3D: phi(x,y,z,t) = cos(omega*t-pi/4)*(x-1)^3*(x+1)^3*(y-1)^3*(y+1)^3
-        #                      *(z-1)^3*(z+1)^3
-        # Not actual solutions, but we can still apply wave operator to them
-        coord_names = ["x", "y", "z"]
-        sym_coords = [pmbl.var(name) for name in coord_names]
+        sym_coords = [pmbl.var(name) for name in ["x", "y", "z"]]
         sym_t = pmbl.var("t")
-        sym_c = pmbl.var("c")
-        sym_omega = pmbl.var("omega")
-        sym_pi = pmbl.var("pi")
-        sym_cos = pmbl.var("cos")
-        sym_phi = sym_cos(sym_omega*sym_t - sym_pi/4)
-        for i in range(dim):
-            sym_phi = sym_phi * (sym_coords[i]-1)**3 * (sym_coords[i]+1)**3
+
+        # f = phi_tt - c^2 * div(grad(phi))
+        sym_f = sym_diff(sym_t)(sym_diff(sym_t)(sym_phi)) - c**2\
+                    * sym_div(sym_grad(dim, sym_phi))
 
         # u = phi_t
         sym_u = sym_diff(sym_t)(sym_phi)
 
         # v = c*grad(phi)
-        sym_v = [sym_c * sym_diff(sym_coords[i])(sym_phi) for i in range(dim)]
+        sym_v = [c * sym_diff(sym_coords[i])(sym_phi) for i in range(dim)]
 
-        # rhs(u part) = c*div(v)
+        # rhs(u part) = c*div(v) + f
         # rhs(v part) = c*grad(u)
         sym_rhs = flat_obj_array(
-                    sym_c * sym_div(sym_v),
-                    make_obj_array([sym_c]) * sym_grad(dim, sym_u))
-
-        c = 2.
-
-        eval_values = {
-            "t": 0.,
-            "c": c,
-            "omega": 1.,
-            "pi": np.pi
-        }
-        eval_values.update({coord_names[i]: nodes[i] for i in range(dim)})
+                    c * sym_div(sym_v) + sym_f,
+                    make_obj_array([c]) * sym_grad(dim, sym_u))
 
-        def sym_eval(expr):
+        def sym_eval(expr, t):
+            coord_names = ["x", "y", "z"]
+            eval_values = {coord_names[i]: nodes[i] for i in range(dim)}
+            eval_values["t"] = t
             return EvaluationMapper(eval_values)(expr)
```

Comment:
- Make this take `**kwargs` instead of just `t`.
- Make it reusable along with #17.

Reply:
Trying to figure out how I could pass in the coordinates via `**kwargs` without requiring different call syntax for each problem dimension... (i.e., `sym_eval(expr, x=nodes[0], y=nodes[1])` for 2D vs. `sym_eval(expr, x=nodes[0], y=nodes[1], z=nodes[2])` for 3D.) Does pymbolic understand array variables? Could I define something along the lines of `pmbl.var('x[0]')`, `pmbl.var('x[1]')`, etc. rather than x,y,z and then do `sym_eval(expr, x=nodes)`?

Reply:
https://documen.tician.de/pymbolic/primitives.html#pymbolic.primitives.make_sym_vector